cpp/src/randomforest/randomforest.cu

```diff
@@ -284,71 +284,40 @@ void print_rf_detailed(const RandomForestMetaData<T, L>* forest) {
 template <class T, class L>
 void build_treelite_forest(ModelHandle* model,
                            const RandomForestMetaData<T, L>* forest,
-                           int num_features, int task_category,
-                           std::vector<unsigned char>& data) {
-  bool check_val = (data).empty();
-  if (not check_val) {
-    // create a temp file
-    const char* filename = std::tmpnam(nullptr);
-    // write the model bytes into the temp file
-    {
-      std::ofstream file(filename, std::ios::binary);
-      file.write((char*)&data[0], data.size());
-    }
-    // read the file as a protobuf model
-    TREELITE_CHECK(TreeliteLoadProtobufModel(filename, model));
+                           int num_features, int task_category) {
+  // Non-zero value here for random forest models.
+  // The value should be set to 0 if the model is gradient boosted trees.
+  int random_forest_flag = 1;
+  ModelBuilderHandle model_builder;
+  // num_output_group is 1 for binary classification and regression
+  // num_output_group is #class for multiclass classification which is the same as task_category
+  int num_output_group = task_category > 2 ? task_category : 1;
+  TREELITE_CHECK(TreeliteCreateModelBuilder(
+    num_features, num_output_group, random_forest_flag, &model_builder));
+
+  if (task_category > 2) {
+    // Multi-class classification
+    TREELITE_CHECK(TreeliteModelBuilderSetModelParam(
+      model_builder, "pred_transform", "max_index"));
   }
 
-  else {
-    // Non-zero value here for random forest models.
-    // The value should be set to 0 if the model is gradient boosted trees.
-    int random_forest_flag = 1;
-    ModelBuilderHandle model_builder;
-    // num_output_group is 1 for binary classification and regression
-    // num_output_group is #class for multiclass classification which is the same as task_category
-    int num_output_group = task_category > 2 ? task_category : 1;
-    TREELITE_CHECK(TreeliteCreateModelBuilder(
-      num_features, num_output_group, random_forest_flag, &model_builder));
-
-    if (task_category > 2) {
-      // Multi-class classification
-      TREELITE_CHECK(TreeliteModelBuilderSetModelParam(
-        model_builder, "pred_transform", "max_index"));
-    }
+  for (int i = 0; i < forest->rf_params.n_trees; i++) {
+    DecisionTree::TreeMetaDataNode<T, L>* tree_ptr = &forest->trees[i];
+    TreeBuilderHandle tree_builder;
 
-    for (int i = 0; i < forest->rf_params.n_trees; i++) {
-      DecisionTree::TreeMetaDataNode<T, L>* tree_ptr = &forest->trees[i];
-      TreeBuilderHandle tree_builder;
+    TREELITE_CHECK(TreeliteCreateTreeBuilder(&tree_builder));
```

Comment:
Does tree_builder need to be freed too?

Reply:
No, since the model builder class frees all trees it contains.